### ISSUE TITLE
How to tracing metrics of keras model with MLFLOW?

### ISSUE THREAD

**Issue by yjiangling** [mlflow/mlflow#13694]
Hi, I'm using keras framework train a model, and tracing the parameters and metrics with MLFLOW like this:

		import mlflow
		import tensorflow as tf
		import mlflow.tensorflow
		mlflow.tensorflow.autolog(every_n_iter=1)

		model = Model()

		train_loss = tf.keras.metrics.Mean(name='train_loss')
		train_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(name='train_accuracy')
		optimizer = tf.keras.optimizers.Adam(1e-4, beta_1=0.9, beta_2=0.98, epsilon=1e-9)
		...

		...

		train_step_signature = [
			tf.TensorSpec(shape=(None, None, 80), dtype=tf.float32),
			tf.TensorSpec(shape=(None), dtype=tf.int32),
			tf.TensorSpec(shape=(None, None), dtype=tf.int32),
			tf.TensorSpec(shape=(None), dtype=tf.int32),
		]
		@tf.function(input_signature=train_step_signature)
		def train_step(xs, xlen, ys, ylen):
			with tf.GradientTape() as tape:
				pred = model.train(xs, xlen, True)
				loss = loss_fun(ys, ylen, pred)

			gradients = tape.gradient(loss, model.trainable_variables)
			# 反向传播
			optimizer.apply_gradients(zip(gradients, model.trainable_variables))

			train_loss(loss)
			train_accuracy(ys, pred2, sample_weight=mask)

		with mlflow.start_run():
			train_dataset = DataLoader(...)

			ckpt = tf.train.Checkpoint(step=tf.Variable(0), model=model, optimizer=optimizer)
			ckpt_manager = tf.train.CheckpointManager(ckpt, checkpoint_dir, max_to_keep=5, keep_checkpoint_every_n_hours=5)

			summary_writer = tf.summary.create_file_writer(log_dir)

			for xs, xlen, ys, ylen in train_dataset:
				train_step(xs, xlen, ys, ylen)
				ckpt.step.assign_add(1)

				if int(ckpt.step) % 50 == 0:
					with summary_writer.as_default():
						tf.summary.scalar('loss', train_loss.result(), step=int(ckpt.step))
						tf.summary.scalar('accuracy', train_accuracy.result(), step=int(ckpt.step))
				...

		

After the training finished, we find the train metrics are corrected record in log files and can be read in TensorBoard, but in MLFLOW, it only trace parameters, the metrics have no data.

![捕获](https://github.com/user-attachments/assets/0d49850b-4387-4252-be8b-61688bf635de)

What's wrong with it? How can I fix it? Anyone can give some help? Many thanks.

**Comment by serena-ruan** [comment]
https://github.com/mlflow/mlflow/blob/master/examples/keras/train.py is an example of using mlflow autologging to auto-log a keras model. Could you provide a reproducible code (that I can run) to validate your issue? `model = Model()` is your model a valid keras model? BTW you can always use `mlflow.log_metrics` to log whatever metrics you want to log

**Comment by yjiangling** [comment]
> https://github.com/mlflow/mlflow/blob/master/examples/keras/train.py is an example of using mlflow autologging to auto-log a keras model. Could you provide a reproducible code (that I can run) to validate your issue? `model = Model()` is your model a valid keras model? BTW you can always use `mlflow.log_metrics` to log whatever metrics you want to log

Oh, I see, yes, **model = Model()** is our model construct with keras framework, and we have trained it many times, but the repository contain quite few scrips, so, is not easy to provide a reproducible code to vildate the issue, I'll try it by myself right now. Thanks a lot for the help.

**Comment by yjiangling** [comment]
> https://github.com/mlflow/mlflow/blob/master/examples/keras/train.py is an example of using mlflow autologging to auto-log a keras model. Could you provide a reproducible code (that I can run) to validate your issue? `model = Model()` is your model a valid keras model? BTW you can always use `mlflow.log_metrics` to log whatever metrics you want to log

You can use this code to to vildate the issue:

		import mlflow
		import numpy as np
		import tensorflow as tf
		import mlflow.tensorflow

		mlflow.tensorflow.autolog(every_n_iter=1)



		class Model(tf.keras.Model):
			def __init__(self, d_model, vocab_size):
				super(Model, self).__init__()
				self.layer1 = tf.keras.layers.Dense(d_model)
				self.layer1 = tf.keras.layers.Dense(vocab_size+1)

			@tf.function
			def train(self, xs):
				outputs = self.layer1(xs)
				logits = self.layers(outputs)
				return logits



		model = Model(512, 1000)

		train_loss = tf.keras.metrics.Mean(name='train_loss')
		optimizer = tf.keras.optimizers.Adam(1e-4, beta_1=0.9, beta_2=0.98, epsilon=1e-9)


		train_step_signature = [
			tf.TensorSpec(shape=(1, None, 80), dtype=tf.float32),
			tf.TensorSpec(shape=(None, None), dtype=tf.int32),
		]
		@tf.function(input_signature=train_step_signature)
		def train_step(xs, ys):
			with tf.GradientTape() as tape:
				logits = model.train(xs)
				xlen = tf.shape(logits)[1][tf.newaxis]
				ylen = tf.shape(ys)[1][tf.newaxis]
				loss = tf.nn.ctc_loss(labels=ys, logits=logits, label_length=ylen, logit_length=xlen, blank_index=-1)

			gradients = tape.gradient(loss, model.trainable_variables)
			# 反向传播
			optimizer.apply_gradients(zip(gradients, model.trainable_variables))

			train_loss(loss)


		with mlflow.start_run():

			ckpt = tf.train.Checkpoint(step=tf.Variable(0), model=model, optimizer=optimizer)
			ckpt_manager = tf.train.CheckpointManager(ckpt, "./checkpoint", max_to_keep=5, keep_checkpoint_every_n_hours=5)

			summary_writer = tf.summary.create_file_writer("./logs")

			for _ in range(2000):
				label_length = np.random.randint(3, 10)
				labels = tf.random.uniform([label_length], minval=1, maxval=1000, dtype=tf.int32)
				input_length = np.random.randint(label_length*3, 200)
				inputs = tf.random.normal([1, input_length, 80], dtype=tf.float32)
				train_step(inputs, labels)
				ckpt.step.assign_add(1)

				if int(ckpt.step) % 50 == 0:
					with summary_writer.as_default():
						tf.summary.scalar('loss', train_loss.result(), step=int(ckpt.step))

				if int(ckpt.step) % 500 == 0:
					ckpt_save_path = ckpt_manager.save()
					logging.info('Save model at {} step to {}'.format(int(ckpt.step), ckpt_save_path))

**Comment by serena-ruan** [comment]
Which tensorflow version are you using? I can't run this code snippet

**Comment by yjiangling** [comment]
> Which tensorflow version are you using? I can't run this code snippet

Sorry for the late reply, there are some bugs in the code and I fix it now, I have tried it on both TensorFlow2.8 and TensorFlow2.12, It's OK now. Could you please try again with the latest code? Thanks a lot

**Comment by serena-ruan** [comment]
Sorry for the late reply.
MLflow tensorflow autologging works by injecting MlflowCallback (`from mlflow.tensorflow.callback import MlflowCallback`) into the existing callbacks, and we only patch `tf.keras.Model's fit method` (https://github.com/mlflow/mlflow/blob/a5b300049af5fed45082d0dd12f65db1ad6be890/mlflow/tensorflow/__init__.py#L1383-L1388), so any other customized functions are not patched. You could inject [MlflowCallbacks](https://github.com/mlflow/mlflow/blob/master/mlflow/tensorflow/callback.py) yourself to enable logging metrics

**Comment by yjiangling** [comment]
> Sorry for the late reply. MLflow tensorflow autologging works by injecting MlflowCallback (`from mlflow.tensorflow.callback import MlflowCallback`) into the existing callbacks, and we only patch `tf.keras.Model's fit method` (
> 
> https://github.com/mlflow/mlflow/blob/a5b300049af5fed45082d0dd12f65db1ad6be890/mlflow/tensorflow/__init__.py#L1383-L1388
> 
> ), so any other customized functions are not patched. You could inject [MlflowCallbacks](https://github.com/mlflow/mlflow/blob/master/mlflow/tensorflow/callback.py) yourself to enable logging metrics

Thanks a lot, I know it. And it seems work with the following code:

`mlflow.log_metric("loss", train_loss.result(), step=int(ckpt.step))`
Since the problem have been solved, I will close the issue right now. Thanks a gain for the helps.
